registry/aws.go

```diff
@@ -0,0 +1,216 @@
+package registry
+
+// References:
+//  - https://github.com/bzon/ecr-k8s-secret-creator
+//  - https://github.com/kubernetes/kubernetes/blob/master/pkg/credentialprovider/aws/aws_credentials.go
+//  - https://github.com/weaveworks/flux/pull/1455
+
+import (
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/ec2metadata"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ecr"
+	"github.com/go-kit/kit/log"
+)
+
+const (
+	// For recognising ECR hosts
+	ecrHostSuffix = ".amazonaws.com"
+	// How long AWS tokens remain valid, according to AWS docs; this
+	// is used as an upper bound, overridden by any sooner expiry
+	// returned in the API response.
+	defaultTokenValid = 12 * time.Hour
+	// how long to skip refreshing a region after we've failed
+	embargoDuration = 10 * time.Minute
+)
+
+type AWSRegistryConfig struct {
+	Regions    []string
+	AccountIDs []string
+}
+
+func contains(strs []string, str string) bool {
+	for _, s := range strs {
+		if s == str {
+			return true
+		}
+	}
+	return false
+}
+
+// ECR registry URLs look like this:
+//
+//     <account-id>.dkr.ecr.<region>.amazonaws.com
+//
+// i.e., they can differ in the account ID and in the region. It's
+// possible to refer to any registry from any cluster (although, being
+// AWS, there will be a cost incurred).
+
+func ImageCredsWithAWSAuth(lookup func() ImageCreds, logger log.Logger, config AWSRegistryConfig) (func() ImageCreds, error) {
+	awsCreds := NoCredentials()
+
+	if len(config.Regions) == 0 {
+		// this forces the AWS SDK to load config, so we can get the default region
+		sess := session.Must(session.NewSessionWithOptions(session.Options{
+			SharedConfigState: session.SharedConfigEnable,
```

Comment:
I think it's a good start, hopefully it will be sufficient long term and we won't need to add too many input variables, as it will always run in a container and won't need to support a host of configuration modes that `eksctl` has to support (environment variables, flags and `~/.aws/config` file with profiles etc).

Reply:
I've tried to leave space for mounting config files (e.g., by attempting to load config before using the ec2 metadata service). You are right that contexts etc. complicate this :-/ We'll see ..